app/scripts/migrations/127.ts

```diff
@@ -0,0 +1,368 @@
+          const duplicateAcrossChains = Object.values(acc)
+            .flatMap((n) => (isObject(n) ? n.rpcEndpoints : []))
+            .find(findDuplicate);
+
+          if (
+            duplicateWithinChain &&
+            network.id === networkState.selectedNetworkClientId
+          ) {
+            // If there's a duplicate RPC url within a chain, and one of the
+            // networks is the globally selected network, prefer to use its network
+            // client id so that `selectedNetworkClientId` can remain unchanged.
+            duplicateWithinChain.networkClientId = network.id;
+          } else if (
+            !duplicateWithinChain &&
+            !duplicateAcrossChains &&
+            typeof network.rpcUrl === 'string' &&
+            isValidUrl(network.rpcUrl)
+          ) {
+            // The endpoint is unique and valid, so add it to the list
+            endpoints.push({
+              networkClientId: network.id,
+              url: network.rpcUrl,
+              type: network.type === 'infura' ? 'infura' : 'custom',
+              ...(network.type !== 'infura' &&
+                typeof network.nickname === 'string' &&
+                network.nickname && { name: network.nickname }),
+            });
+          }
+        }
+        return endpoints;
+      },
+      [],
+    );
+
+    // If there were no valid unique endpoints, then omit the network
+    // configuration for this chain id. The network controller requires
+    // configurations to have at least 1 endpoint.
+    if (rpcEndpoints.length === 0) {
+      return acc;
+    }
+
+    // Use the tie breaker network as the default rpc endpoint
+    const defaultRpcEndpointIndex = Math.max(
+      rpcEndpoints.findIndex(
+        (endpoint) => endpoint.networkClientId === tieBreaker?.id,
+      ),
+      // Or arbitrarily default to the first endpoint if we don't have a tie breaker
+      0,
+    );
+
+    // Calculate the unique array of non-empty block explorer urls
+    const blockExplorerUrls = [
+      ...networks.reduce((urls, network) => {
+        if (
+          isObject(network.rpcPrefs) &&
+          typeof network.rpcPrefs.blockExplorerUrl === 'string' &&
+          network.rpcPrefs.blockExplorerUrl
+        ) {
+          urls.add(network.rpcPrefs.blockExplorerUrl);
+        }
+        return urls;
+      }, new Set()),
+    ];
+
+    // Use the tie breaker network as the default block explorer, if it has one
+    const defaultBlockExplorerUrlIndex =
+      blockExplorerUrls.length === 0
+        ? undefined
+        : Math.max(
+            blockExplorerUrls.findIndex(
+              (url) =>
+                isObject(tieBreaker?.rpcPrefs) &&
+                url === tieBreaker.rpcPrefs.blockExplorerUrl,
+            ),
+            // Or arbitrarily default to the first url
+            0,
+          );
+
+    // Use the cononical network name and currency, if we have constants for them.
```

Comment:
It would usually be bad to reference product code in a migration.  But it seemed useful here.

If you have networks like 'quicknode arbitrum' and 'llama nodes arbitrum', neither makes sense to become the overall network name post-merge.  Those old network names do become RPC endpoint names.  But the overall network should just 'arbitrum' now.

This could be relaxed to only go cononical if there are multiple networks per chain.